### ISSUE TITLE
DIM extension keeps disappearing

### ISSUE THREAD

**Issue by negativeExponent** [libretro/RetroArch#14425]
# First and foremost consider this:
- Only RetroArch bugs should be filed here. Not core bugs or game bugs
- This is not a forum or a help section, this is strictly developer oriented

## Description

As stated in title, DIM extension keeps disappearing from file browser

### Expected behavior

DIM extension, which is a valid extension for floppy disk format keeps disappearing from file browser.

### Actual behavior

DIM sometimes appear the first time you enter the folder with this file extension, but will just disappear after... or they wont appear at all.

### Steps to reproduce the bug

* updated assets
* updates core info
* updated px68k (or probably any other core supporting .dim file extension)
* (do i need to upgrade my computer as well?)

1. open RA then browse to folder with files in DIM extension.
2. either they appear or will disappear after entering the folder a few times.
3. and so on...

### Bisect Results

[Try to bisect and tell us when this started happening]

### Version/Commit
You can find this information under Information/System Information

- RetroArch: latest commit/master branch

### Environment information

- OS: arch
- Compiler: gcc 12

"what do i need to 'reset' this time?"

**Comment by gouchi** [comment]
I can't reproduce the issue using commit [e8f4425](https://github.com/libretro/RetroArch/commit/e8f4425eedec07c7d3c7d7f906c029dcfd4aa36a) and latest core info files.

I have those options for the file browser

![file-browser-options](https://user-images.githubusercontent.com/7869503/190899168-a8366981-83d5-4a31-b66d-a3829ab3b0e0.png)

**Comment by bslenul** [comment]
I can reproduce but only with RA built with `DEBUG=1` for whatever reason...:

https://user-images.githubusercontent.com/33353403/190903706-6946fed2-1173-4459-9a2e-2a08a3d599a2.mp4

1. I show that ONLY PX68k core is installed.
2. I navigate to a test folder with a few .dim and .m3u files, the .dim files don't appear, the .m3u do.
3. I download FBNeo core and for some reason it makes the .dim files to appear (FBNeo doesn't support .dim).
4. I delete FBNeo and the files disappear.
5. Finally I start a non-debug version of RA (same commit) and no issue there, .dim files appear directly.

I first thought it was some core info cache shenanigans but nope, it happens with it disabled as well. Similar issue with Flycast and .chd files.

**Comment by negativeExponent** [comment]
just i've described above, the DIM may appear a few times after just launching RA, but eventually it will disappear. set some commits bad coz i've thought found the problematic commit, but yeah, after entering the directory, file just disappears

installing fbneo does nothing on my end.

**Comment by bslenul** [comment]
OK, my issue above might be unrelated then. Tested on my Linux Mint VM out of curiosity and I can reproduce with the nightly version from the PPA:

![image](https://user-images.githubusercontent.com/33353403/190912334-faf54643-9699-431d-9e65-4bb74e4427ca.png)

**Comment by LibretroAdmin** [comment]
I will attempt to look into this.

**Comment by LibretroAdmin** [comment]
> just i've described above, the DIM may appear a few times after just launching RA, but eventually it will disappear. set some commits bad coz i've thought found the problematic commit, but yeah, after entering the directory, file just disappears
> 
> installing fbneo does nothing on my end.
> 
> there's a certain commit that seems OK for the most part... but everything after that is problematic [200897f](https://github.com/libretro/RetroArch/commit/200897f61e88f8f2fc6e0f5f9215f9a33efed970)

Can you confirm the issue starts happening after that commit? Can we take this as bisect results?

**Comment by bslenul** [comment]
It was a PITA to bisect due to random behaviors but it looks like 88c2ae9ddfb421b2d4eb1493d60dc4aeca502c0e is the guilty commit, with the previous commit (acfe990ec6e322377936fe06d7dd29274c23bbd3) I can't reproduce the issue at all.

**Comment by bslenul** [comment]
Might need more testing (and maybe a confirmation from negativeExponent?) but it seems to work fine if I just restore `newstring[0]                     = '\0';` from "menu_cbs_deferred_push.c" (tested on Win10 and my Linux Mint VM):
```diff
diff --git a/menu/cbs/menu_cbs_deferred_push.c b/menu/cbs/menu_cbs_deferred_push.c
index 27c427c2ae..6c3f032302 100644
--- a/menu/cbs/menu_cbs_deferred_push.c
+++ b/menu/cbs/menu_cbs_deferred_push.c
@@ -516,6 +516,7 @@ static int general_push(menu_displaylist_info_t *info,
             struct retro_system_info *system =
                &runloop_state_get_ptr()->system.info;

+            newstring[0]                     = '\0';
             attr.i                           = 0;

             string_list_initialize(&str_list2);
```

**Comment by LibretroAdmin** [comment]
Yeah, `string_list_join_concat` doesn't use strlcpy under the hood but strlcat in a loop. So I think it's prudent there to NULL terminate before we hand the string over to that function.

**Comment by negativeExponent** [comment]
> Might need more testing (and maybe a confirmation from negativeExponent?) but it seems to work fine if I just restore `newstring[0] = '\0';` from "menu_cbs_deferred_push.c" (tested on Win10 and my Linux Mint VM):
> 
> ```diff
> diff --git a/menu/cbs/menu_cbs_deferred_push.c b/menu/cbs/menu_cbs_deferred_push.c
> index 27c427c2ae..6c3f032302 100644
> --- a/menu/cbs/menu_cbs_deferred_push.c
> +++ b/menu/cbs/menu_cbs_deferred_push.c
> @@ -516,6 +516,7 @@ static int general_push(menu_displaylist_info_t *info,
>              struct retro_system_info *system =
>                 &runloop_state_get_ptr()->system.info;
> 
> +            newstring[0]                     = '\0';
>              attr.i                           = 0;
> 
>              string_list_initialize(&str_list2);
> ```

i can confirm that just adding this line fixes the issue... feel free to close thread when this gets merged. ill reopen if needed

**Comment by LibretroAdmin** [comment]
Awesome guys, thanks so much for investigating this and my bad for the issue. We'll gather up a few more bug fixes and release a 11.1.2 sometime later this week, will be good to squash a few more bugs (not the regression sort but longstanding issues).

**Comment by LibretroAdmin** [comment]
I've had a look at the other  `string_list_join_concat` calls in the codebase and thankfully it seems we are safe there, strings passed are properly NULL-terminated. It seemed to be only this spot in menu_cbs_deferred_push.c where there was an actual mistake.

**Comment by LibretroAdmin** [comment]
https://github.com/libretro/RetroArch/commit/b2521c53ae1314924fbb7921ff37802a22c6febf

Fixed. Please confirm one last time this resolves it and then go ahead and close the issue.

**Comment by negativeExponent** [comment]
yeah, its working now. tnx
